File: src/translator/zh2en/provider/requirement.ts
import { DATA as POE_DATA } from "../../../data/poe/index.js";
import type {
    Requirement,
    RequirementSuffix,
} from "../../../data/poe/types.js";

export class RequirementProvider {
    private readonly zhIdx = new Map<string, Requirement>();
    private readonly suffixesZhIdx = new Map<string, RequirementSuffix>();

    constructor() {
        for (const r of POE_DATA.requirements) {
            const zh = r.zh;
            this.zhIdx.set(zh, r);
        }

        for (const s of POE_DATA.requirementSuffixes) {
            const zh = s.zh;
            this.suffixesZhIdx.set(zh, s);
        }
    }

    provideByZh(zh: string): Requirement | undefined {
        return this.zhIdx.get(zh);
    }

    provideSuffixByZh(zh: string): RequirementSuffix | undefined {
        return this.suffixesZhIdx.get(zh);
    }
}

File: go/translator/zh2en/provider/requirement.go
package provider

import (
	"github.com/cn-poe-community/cn-poe-utils/go/data/poe"
)

type RequirementProvider struct {
	zhIdx       map[string]*poe.Requirement
	suffixZhIdx map[string]*poe.RequirementSuffix
}

func NewRequirementProvider() *RequirementProvider {
	zhIdx := make(map[string]*poe.Requirement)
	suffixZhIdx := make(map[string]*poe.RequirementSuffix)

	for i := range poe.DATA.Requirements {
		req := &poe.DATA.Requirements[i]
		zhIdx[req.Zh] = req
	}
	for i := range poe.DATA.RequirementSuffixes {
		suffix := &poe.DATA.RequirementSuffixes[i]
		suffixZhIdx[suffix.Zh] = suffix
	}

	return &RequirementProvider{
		zhIdx:       zhIdx,
		suffixZhIdx: suffixZhIdx,
	}
}

func (p *RequirementProvider) ProvideByZh(zh string) *poe.Requirement {
	if req, ok := p.zhIdx[zh]; ok {
		return req
	}
	return nil
}

func (p *RequirementProvider) ProvideSuffixByZh(zh string) *poe.RequirementSuffix {
	if suffix, ok := p.suffixZhIdx[zh]; ok {
		return suffix
	}
	return nil
}
